### Тестовое задание
Ниже выполнено тестовое задание от некоторого государственного бюджетного учреждения. В рассматриваемом датафрейме содержатся данные о проведении медицинских исследований в ряде медицинских учреждений Москвы. Задача заключается в написании скрипта обработки данных и выгрузки отчета по определенной форме. Данные необходимо было импортировать из файла Excel, создать отчет, выгрузить его в Excel и на его основе создать дашборд в Datalens. 

### Подготовка данных

In [1]:
# Импорт библиотек:

import pandas as pd
import numpy as np
import datetime as dt
pd.options.mode.chained_assignment = None

In [2]:
# Сохраним в переменную датафрейм из файла Excel:

data = pd.read_excel('/mnt/HC_Volume_18315164/home-jupyter/jupyter-d-kirilovichev/Приложения.xlsx')
display(data.head())
print(data.info())

ID_MO_EMIAS DISTRICT         AE_TITLE DEVICE_TYPE  \
0  ГБУЗ "ГКБ № 52 ДЗМ"     СЗАО  AQUILION1_GKB52          КТ   
1  ГБУЗ "ГКБ № 52 ДЗМ"     СЗАО  AQUILION1_GKB52          КТ   
2   ГБУЗ "ГП № 19 ДЗМ"     ЮВАО    EXCELMRI_GP19         МРТ   
3   ГБУЗ "ГП № 19 ДЗМ"     ЮВАО    EXCELMRI_GP19         МРТ   
4   ГБУЗ "ГП № 19 ДЗМ"     ЮВАО    EXCELMRI_GP19         МРТ   

                                           STUDY_UID SERVICE_TYPE BODY_PART  \
0  1.2.40.0.13.1.31663082429482479615925629310980...           КТ   Abdomen   
1  1.2.40.0.13.1.54587529764833093390270523355606...           КТ     Chest   
2  1.2.40.0.13.1.83400150319274737474751731880019...          МРТ    Pelvis   
3  1.2.40.0.13.1.19280153773959545385899576777341...          МРТ    Pelvis   
4  1.2.40.0.13.1.12150928157332376871218930449508...          МРТ    Pelvis   

   PROCEDURE_CODE                                     PROCEDURE_NAME  \
0             103  Компьютерная томография органов брюшной полост...   
1              97     Компьютерная томография органов грудной клетки   
2             370  Мультипараметрическая магнитно-резонансная том...   
3             370  Мультипараметрическая магнитно-резонансная том...   
4             370  Мультипараметрическая магнитно-резонансная том...   

        STUDY_DATETIME  ...      REPORT_CREATION    LAST_VALIDATION_1  \
0  30.11.2022 15:36:00  ...  30.11.2022 16:25:00  30.11.2022 16:42:19   
1  28.11.2022 07:26:00  ...  28.11.2022 07:40:00  28.11.2022 07:44:46   
2  28.11.2022 12:21:00  ...  01.12.2022 11:12:00  01.12.2022 11:46:29   
3  01.12.2022 11:37:00  ...  01.12.2022 16:01:00  01.12.2022 16:27:16   
4  29.11.2022 11:42:00  ...  29.11.2022 19:04:00  29.11.2022 19:29:35   

     LAST_VALIDATION_2 DOCTOR_CODE_1  \
0  01.12.2022 14:56:31      CNTCT417   
1  28.11.2022 11:10:49    CNTCT10011   
2  01.12.2022 21:19:41      CNTCT439   
3  01.12.2022 21:29:10      CNTCT439   
4  30.11.2022 11:57:57      CNTCT439   

                                         DESCRIPTION  \
0  \nГрыж на передней брюшной стенке не выявлено....   
1  \nВ легких определяются немногочисленные ретик...   
2  ПЕРВОЕ ЧТЕНИЕ: врач-рентгенолог Воронин  Алекс...   
3  ПЕРВОЕ ЧТЕНИЕ: врач-рентгенолог Воронин  Алекс...   
4  ПЕРВОЕ ЧТЕНИЕ: врач-рентгенолог Воронин  Алекс...   

                                          CONCLUSION PNM_SEVERITY_GDT  \
0  Дополнительное заключение\nПравое паховое коль...              NaN   
1  Дополнительное заключение\nПрицельно осмотрены...              NaN   
2  ПЕРВОЕ ЧТЕНИЕ: врач-рентгенолог Воронин  Алекс...              NaN   
3  ПЕРВОЕ ЧТЕНИЕ: врач-рентгенолог Воронин  Алекс...              NaN   
4  ПЕРВОЕ ЧТЕНИЕ: врач-рентгенолог Воронин  Алекс...              NaN   

  CLINTAGS_TUB_VALUE PNM_SEVERITY_ADT RECOMMENDATION  
0                NaN              NaN            NaN  
1                NaN              NaN            NaN  
2                NaN              NaN            NaN  
3                NaN              NaN            NaN  
4                NaN              NaN            NaN  

[5 rows x 22 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 179131 entries, 0 to 179130
Data columns (total 22 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   ID_MO_EMIAS         179131 non-null  object 
 1   DISTRICT            179131 non-null  object 
 2   AE_TITLE            179131 non-null  object 
 3   DEVICE_TYPE         179131 non-null  object 
 4   STUDY_UID           179131 non-null  object 
 5   SERVICE_TYPE        179131 non-null  object 
 6   BODY_PART           179131 non-null  object 
 7   PROCEDURE_CODE      179131 non-null  int64  
 8   PROCEDURE_NAME      179131 non-null  object 
 9   STUDY_DATETIME      179131 non-null  object 
 10  STUDY_CREATEDTIME   179131 non-null  object 
 11  FIN_PRICE           179131 non-null  object 
 12  REPORT_CREATION     179074 non-null  object 
 13  LAST_VALIDATION_1   179131 non-null  object 
 14  LAST_VALIDATION_2   168718 non-null  object 
 15  DOCTOR_CODE_1       178498 non-nul

<span style="color:blue">Библиотеки и данные импортированы.</span>

### Подготовка отчета для создания дашборда

In [3]:
# 1. Отберем все записи по процедуре 'Компьютерная томография органов грудной клетки, 
# брюшной полости и малого таза с контрастированием':

data1 = data[data['PROCEDURE_NAME']  == 
             'Компьютерная томография органов грудной клетки, брюшной полости и малого таза с контрастированием']
display(data1.head())
print(data1.info())

ID_MO_EMIAS DISTRICT       AE_TITLE DEVICE_TYPE  \
47   ГБУЗ "ГКБ им. С.П. Боткина ДЗМ"      САО   AQUILION_BOT          КТ   
284               ГБУЗ "ГВВ № 3 ДЗМ"     СВАО  AQUILION_GVV3          КТ   
285               ГБУЗ "ГВВ № 3 ДЗМ"     СВАО  AQUILION_GVV3          КТ   
286               ГБУЗ "ГВВ № 3 ДЗМ"     СВАО  AQUILION_GVV3          КТ   
287               ГБУЗ "ГВВ № 3 ДЗМ"     СВАО  AQUILION_GVV3          КТ   

                                             STUDY_UID SERVICE_TYPE  \
47   1.2.40.0.13.1.24625127896634081911387737364213...           КТ   
284  1.2.40.0.13.1.30418767661311803934198222959245...           КТ   
285  1.2.40.0.13.1.21804036516467810836637169435609...           КТ   
286  1.2.40.0.13.1.15232219916349159455333183362956...           КТ   
287  1.2.40.0.13.1.14607635274410420056808620437745...           КТ   

     BODY_PART  PROCEDURE_CODE  \
47   WHOLEBODY             404   
284  WHOLEBODY             404   
285  WHOLEBODY             404   
286  WHOLEBODY             404   
287  WHOLEBODY             404   

                                        PROCEDURE_NAME       STUDY_DATETIME  \
47   Компьютерная томография органов грудной клетки...  30.11.2022 15:08:00   
284  Компьютерная томография органов грудной клетки...  30.11.2022 09:40:00   
285  Компьютерная томография органов грудной клетки...  30.11.2022 14:55:00   
286  Компьютерная томография органов грудной клетки...  30.11.2022 14:32:00   
287  Компьютерная томография органов грудной клетки...  30.11.2022 14:11:00   

     ...      REPORT_CREATION    LAST_VALIDATION_1    LAST_VALIDATION_2  \
47   ...  30.11.2022 16:52:00  30.11.2022 16:53:00  01.12.2022 13:03:17   
284  ...  30.11.2022 10:21:00  30.11.2022 10:50:42  30.11.2022 10:50:42   
285  ...  30.11.2022 19:17:00  30.11.2022 19:38:09  30.11.2022 19:38:09   
286  ...  30.11.2022 18:49:00  30.11.2022 19:09:52  30.11.2022 19:09:52   
287  ...  30.11.2022 18:29:00  30.11.2022 18:41:20  30.11.2022 18:41:20   

    DOCTOR_CODE_1                                        DESCRIPTION  \
47     CNTCT91803  Исследование выполнено по программе мультиспир...   
284     CNTCT9801  \nКлиническая информация: С18.2.  злокачествен...   
285     CNTCT9801  \nКод МКБ: С62.1. семинома правого яичка, мтс ...   
286     CNTCT9801  \nКлиническая информация: R59.9.\nКТ ОРГАНОВ Г...   
287     CNTCT9801  \nКлиническая информация: N28.1. Цель исследов...   

                                            CONCLUSION PNM_SEVERITY_GDT  \
47   КТ-данных за прогрессирование основного заболе...              NaN   
284  \nВ легких без вторичных изменений, при сравне...              NaN   
285  \nПо сравнению с КТ-исследованием  от 19.04.20...              NaN   
286  \nКТ - картина мелких очагов уплотнения в легк...              NaN   
287  \nВ легких без очаговых изменений.Косвенные пр...              NaN   

    CLINTAGS_TUB_VALUE PNM_SEVERITY_ADT  \
47                 NaN              NaN   
284                NaN              NaN   
285                NaN              NaN   
286                NaN              NaN   
287                NaN              NaN   

                                        RECOMMENDATION  
47                                                 NaN  
284  \nИсследование выполнено при пограничных данны...  
285                                                NaN  
286                                                NaN  
287                                                NaN  

[5 rows x 22 columns]

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2235 entries, 47 to 178576
Data columns (total 22 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID_MO_EMIAS         2235 non-null   object 
 1   DISTRICT            2235 non-null   object 
 2   AE_TITLE            2235 non-null   object 
 3   DEVICE_TYPE         2235 non-null   object 
 4   STUDY_UID           2235 non-null   object 
 5   SERVICE_TYPE        2235 non-null   object 
 6   BODY_PART           2235 non-null   object 
 7   PROCEDURE_CODE      2235 non-null   int64  
 8   PROCEDURE_NAME      2235 non-null   object 
 9   STUDY_DATETIME      2235 non-null   object 
 10  STUDY_CREATEDTIME   2235 non-null   object 
 11  FIN_PRICE           2235 non-null   object 
 12  REPORT_CREATION     2221 non-null   object 
 13  LAST_VALIDATION_1   2235 non-null   object 
 14  LAST_VALIDATION_2   2157 non-null   object 
 15  DOCTOR_CODE_1       2235 non-null   object 
 16  DES

<span style="color:blue">Нужные записи отобраны, их осталось 2235 строк.</span>

In [4]:
# 2. Отберем строки в столбце conclusion, по которым есть ключевые слова "recist" и "тип ответа":

data2 = data1[data1['CONCLUSION'].str.lower().str.contains('recist|тип ответа', na = False)]
data2 = data2.reset_index(drop = True)
display(data2.head())
print(data2.info())

ID_MO_EMIAS            DISTRICT          AE_TITLE  \
0     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   
1     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   
2              ГБУЗ "ГКОБ № 1 ДЗМ"                 ЦАО    AQUILION_GKOB1   
3  ГБУЗ "ГКОБ № 1 ДЗМ", Филиал № 1                ЮЗАО  AQUILION_GKOB1F1   
4     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   

  DEVICE_TYPE                                          STUDY_UID SERVICE_TYPE  \
0          КТ  1.2.840.113619.2.428.3.1930040987.340.16698629...           КТ   
1          КТ  1.2.840.113619.2.428.3.1930040987.340.16698629...           КТ   
2          КТ  1.2.40.0.13.1.20397098119916846422725270055549...           КТ   
3          КТ  1.2.40.0.13.1.31668921814789388389199005221429...           КТ   
4          КТ  1.2.840.113619.2.428.3.1930040987.288.16696088...           КТ   

   BODY_PART  PROCEDURE_CODE  \
0  WHOLEBODY             404   
1  WHOLEBODY             404   
2  WHOLEBODY             404   
3  WHOLEBODY             404   
4  WHOLEBODY             404   

                                      PROCEDURE_NAME       STUDY_DATETIME  \
0  Компьютерная томография органов грудной клетки...  02.12.2022 13:04:00   
1  Компьютерная томография органов грудной клетки...  02.12.2022 08:38:00   
2  Компьютерная томография органов грудной клетки...  04.12.2022 11:50:00   
3  Компьютерная томография органов грудной клетки...  04.12.2022 09:15:00   
4  Компьютерная томография органов грудной клетки...  30.11.2022 13:13:00   

   ...      REPORT_CREATION    LAST_VALIDATION_1    LAST_VALIDATION_2  \
0  ...  02.12.2022 15:21:00  02.12.2022 16:43:58  02.12.2022 16:43:58   
1  ...  02.12.2022 11:15:00  02.12.2022 12:06:38  03.12.2022 13:26:21   
2  ...  04.12.2022 13:07:00  04.12.2022 13:45:48  04.12.2022 13:45:48   
3  ...  04.12.2022 09:50:00  04.12.2022 11:14:33  04.12.2022 13:47:20   
4  ...  30.11.2022 20:24:00  30.11.2022 20:44:24  30.11.2022 20:44:24   

  DOCTOR_CODE_1                                        DESCRIPTION  \
0     CNTCT1278  \nПри сравнении с исследованием от 29/09/22 - ...   
1     CNTCT1278  \nИсследование повторное, предыдущее от 27/08/...   
2     CNTCT1284  \nСравнение с 18.10.22:\n\nКлинические данные:...   
3     CNTCT1284  \nСравнение с 22.10.22:\n\nКлинические данные:...   
4      CNTCT116  \nВ сравнении с КТ исследованием от 23.08.22г ...   

                                          CONCLUSION PNM_SEVERITY_GDT  \
0  \nТип ответа по recist: нельзя оценить\n\nПри ...              NaN   
1  \nИзмеримых очагов для оценки по Recist 1.1 не...              NaN   
2  \nКТ-картина очаговых изменений легких вторичн...              NaN   
3  \nКТ-картина опухоли левой молочной железы, вт...              NaN   
4  \nКТ-картина центральной опухоли верхней доли ...              NaN   

  CLINTAGS_TUB_VALUE PNM_SEVERITY_ADT RECOMMENDATION  
0                NaN              NaN            NaN  
1                NaN              NaN            NaN  
2                NaN              NaN            NaN  
3                NaN              NaN            NaN  
4                NaN              NaN            NaN  

[5 rows x 22 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 179 entries, 0 to 178
Data columns (total 22 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID_MO_EMIAS         179 non-null    object 
 1   DISTRICT            179 non-null    object 
 2   AE_TITLE            179 non-null    object 
 3   DEVICE_TYPE         179 non-null    object 
 4   STUDY_UID           179 non-null    object 
 5   SERVICE_TYPE        179 non-null    object 
 6   BODY_PART           179 non-null    object 
 7   PROCEDURE_CODE      179 non-null    int64  
 8   PROCEDURE_NAME      179 non-null    object 
 9   STUDY_DATETIME      179 non-null    object 
 10  STUDY_CREATEDTIME   179 non-null    object 
 11  FIN_PRICE           179 non-null    object 
 12  REPORT_CREATION     179 non-null    object 
 13  LAST_VALIDATION_1   179 non-null    object 
 14  LAST_VALIDATION_2   179 non-null    object 
 15  DOCTOR_CODE_1       179 non-null    object 
 16  DESCRIPT

<span style="color:blue">Нужные строки по заданным ключевым словам отобраны, всего их 179.</span>

In [5]:
# 3. Создадим столбец keyword и запишем отобранные ключевые слова "recist" и "тип ответа":

# создадим пустой столбец keyword
data2['keyword'] = ''

# присвоим строкам в столбце keyword значение "Тип ответа", где "тип ответа" встречается в столбце conclusion
for i in range (data2.shape[0]):
    if 'тип ответа' in data2['CONCLUSION'][i].lower():
        data2['keyword'][i] = 'Тип ответа'
print()

# присвоим строкам в столбце keyword значение "RECIST", где "recist" встречается в столбце conclusion
for j in range (data2.shape[0]):
    if 'recist' in data2['CONCLUSION'][j].lower():
        data2['keyword'][j] = 'RECIST'
print(data2['keyword'].value_counts())
display(data2.head())


RECIST        163
Тип ответа     16
Name: keyword, dtype: int64


ID_MO_EMIAS            DISTRICT          AE_TITLE  \
0     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   
1     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   
2              ГБУЗ "ГКОБ № 1 ДЗМ"                 ЦАО    AQUILION_GKOB1   
3  ГБУЗ "ГКОБ № 1 ДЗМ", Филиал № 1                ЮЗАО  AQUILION_GKOB1F1   
4     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   

  DEVICE_TYPE                                          STUDY_UID SERVICE_TYPE  \
0          КТ  1.2.840.113619.2.428.3.1930040987.340.16698629...           КТ   
1          КТ  1.2.840.113619.2.428.3.1930040987.340.16698629...           КТ   
2          КТ  1.2.40.0.13.1.20397098119916846422725270055549...           КТ   
3          КТ  1.2.40.0.13.1.31668921814789388389199005221429...           КТ   
4          КТ  1.2.840.113619.2.428.3.1930040987.288.16696088...           КТ   

   BODY_PART  PROCEDURE_CODE  \
0  WHOLEBODY             404   
1  WHOLEBODY             404   
2  WHOLEBODY             404   
3  WHOLEBODY             404   
4  WHOLEBODY             404   

                                      PROCEDURE_NAME       STUDY_DATETIME  \
0  Компьютерная томография органов грудной клетки...  02.12.2022 13:04:00   
1  Компьютерная томография органов грудной клетки...  02.12.2022 08:38:00   
2  Компьютерная томография органов грудной клетки...  04.12.2022 11:50:00   
3  Компьютерная томография органов грудной клетки...  04.12.2022 09:15:00   
4  Компьютерная томография органов грудной клетки...  30.11.2022 13:13:00   

   ...    LAST_VALIDATION_1    LAST_VALIDATION_2 DOCTOR_CODE_1  \
0  ...  02.12.2022 16:43:58  02.12.2022 16:43:58     CNTCT1278   
1  ...  02.12.2022 12:06:38  03.12.2022 13:26:21     CNTCT1278   
2  ...  04.12.2022 13:45:48  04.12.2022 13:45:48     CNTCT1284   
3  ...  04.12.2022 11:14:33  04.12.2022 13:47:20     CNTCT1284   
4  ...  30.11.2022 20:44:24  30.11.2022 20:44:24      CNTCT116   

                                         DESCRIPTION  \
0  \nПри сравнении с исследованием от 29/09/22 - ...   
1  \nИсследование повторное, предыдущее от 27/08/...   
2  \nСравнение с 18.10.22:\n\nКлинические данные:...   
3  \nСравнение с 22.10.22:\n\nКлинические данные:...   
4  \nВ сравнении с КТ исследованием от 23.08.22г ...   

                                          CONCLUSION PNM_SEVERITY_GDT  \
0  \nТип ответа по recist: нельзя оценить\n\nПри ...              NaN   
1  \nИзмеримых очагов для оценки по Recist 1.1 не...              NaN   
2  \nКТ-картина очаговых изменений легких вторичн...              NaN   
3  \nКТ-картина опухоли левой молочной железы, вт...              NaN   
4  \nКТ-картина центральной опухоли верхней доли ...              NaN   

  CLINTAGS_TUB_VALUE PNM_SEVERITY_ADT RECOMMENDATION keyword  
0                NaN              NaN            NaN  RECIST  
1                NaN              NaN            NaN  RECIST  
2                NaN              NaN            NaN  RECIST  
3                NaN              NaN            NaN  RECIST  
4                NaN              NaN            NaN  RECIST  

[5 rows x 23 columns]

<span style="color:blue">Итого в столбце keyword 163 строкам присвоено значение "RECIST" и 16 строкам присвоено значение "Тип ответа".</span>

In [6]:
# 4. Создадим столбец estimation_Recist, в который запишем заключение из поля CONCLUSION по ключевым словам:

# создадим список ключевых слов, по которым будем делать поиск
types_answer = ['стабилизация', 'прогрессирование', 'частичный ответ', 'полный ответ']

# создадим список с ответами, которые будет вписывать в столбец estimation_Recist
types_answer_full = ['стабилизация (Stable Disease)', 'прогрессирование (Progressive Disease)', 
                     'частичный ответ (Partial Response)', 'полный ответ (Complete Response)']

# создадим пустой столбец estimation_Recist
data2['estimation_Recist'] = ''

# для строк, где встречаются ключевые слова, впишем ответы
for i in range (data2.shape[0]):
    for j in range(len(types_answer)):
        if types_answer[j] in data2['CONCLUSION'][i].lower():
            data2['estimation_Recist'][i] = types_answer_full[j]
display(data2.head())
data2['estimation_Recist'].value_counts()

ID_MO_EMIAS            DISTRICT          AE_TITLE  \
0     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   
1     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   
2              ГБУЗ "ГКОБ № 1 ДЗМ"                 ЦАО    AQUILION_GKOB1   
3  ГБУЗ "ГКОБ № 1 ДЗМ", Филиал № 1                ЮЗАО  AQUILION_GKOB1F1   
4     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   

  DEVICE_TYPE                                          STUDY_UID SERVICE_TYPE  \
0          КТ  1.2.840.113619.2.428.3.1930040987.340.16698629...           КТ   
1          КТ  1.2.840.113619.2.428.3.1930040987.340.16698629...           КТ   
2          КТ  1.2.40.0.13.1.20397098119916846422725270055549...           КТ   
3          КТ  1.2.40.0.13.1.31668921814789388389199005221429...           КТ   
4          КТ  1.2.840.113619.2.428.3.1930040987.288.16696088...           КТ   

   BODY_PART  PROCEDURE_CODE  \
0  WHOLEBODY             404   
1  WHOLEBODY             404   
2  WHOLEBODY             404   
3  WHOLEBODY             404   
4  WHOLEBODY             404   

                                      PROCEDURE_NAME       STUDY_DATETIME  \
0  Компьютерная томография органов грудной клетки...  02.12.2022 13:04:00   
1  Компьютерная томография органов грудной клетки...  02.12.2022 08:38:00   
2  Компьютерная томография органов грудной клетки...  04.12.2022 11:50:00   
3  Компьютерная томография органов грудной клетки...  04.12.2022 09:15:00   
4  Компьютерная томография органов грудной клетки...  30.11.2022 13:13:00   

   ...    LAST_VALIDATION_2 DOCTOR_CODE_1  \
0  ...  02.12.2022 16:43:58     CNTCT1278   
1  ...  03.12.2022 13:26:21     CNTCT1278   
2  ...  04.12.2022 13:45:48     CNTCT1284   
3  ...  04.12.2022 13:47:20     CNTCT1284   
4  ...  30.11.2022 20:44:24      CNTCT116   

                                         DESCRIPTION  \
0  \nПри сравнении с исследованием от 29/09/22 - ...   
1  \nИсследование повторное, предыдущее от 27/08/...   
2  \nСравнение с 18.10.22:\n\nКлинические данные:...   
3  \nСравнение с 22.10.22:\n\nКлинические данные:...   
4  \nВ сравнении с КТ исследованием от 23.08.22г ...   

                                          CONCLUSION PNM_SEVERITY_GDT  \
0  \nТип ответа по recist: нельзя оценить\n\nПри ...              NaN   
1  \nИзмеримых очагов для оценки по Recist 1.1 не...              NaN   
2  \nКТ-картина очаговых изменений легких вторичн...              NaN   
3  \nКТ-картина опухоли левой молочной железы, вт...              NaN   
4  \nКТ-картина центральной опухоли верхней доли ...              NaN   

  CLINTAGS_TUB_VALUE PNM_SEVERITY_ADT RECOMMENDATION keyword  \
0                NaN              NaN            NaN  RECIST   
1                NaN              NaN            NaN  RECIST   
2                NaN              NaN            NaN  RECIST   
3                NaN              NaN            NaN  RECIST   
4                NaN              NaN            NaN  RECIST   

                        estimation_Recist  
0                                          
1                                          
2           стабилизация (Stable Disease)  
3  прогрессирование (Progressive Disease)  
4           стабилизация (Stable Disease)  

[5 rows x 24 columns]

стабилизация (Stable Disease)             70
прогрессирование (Progressive Disease)    43
                                          42
частичный ответ (Partial Response)        20
полный ответ (Complete Response)           4
Name: estimation_Recist, dtype: int64

<span style="color:blue">Столбец estimation_Recist заполнен значениями из поля CONCLUSION, при этом 42 строки остались пустыми.</span>

In [7]:
# 5. Оставим запись в поле «estimation_Recist» пустой, если в CONCLUSION имеются записи, 
# которые указывают на невозможность оценить тип ответа по Recist: 

# создадим список с "отказными" записями
score_unavailable = ['оценка по recist невозможна', 'baseline не указан', 
                     'по критериям recist не выявлены', 'не попадают под критерии recist', 
                     'данные nadir и baseline не указаны']

# сделаем пустыми строки, по которым в столбце conclusion встречаются "отказные" записи
for i in range (data2.shape[0]):
    for j in range(len(score_unavailable)):
        if score_unavailable[j] in data2['CONCLUSION'][i].lower():
            data2['estimation_Recist'][i] = ''
            
display(data2.head())
data2['estimation_Recist'].value_counts()

ID_MO_EMIAS            DISTRICT          AE_TITLE  \
0     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   
1     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   
2              ГБУЗ "ГКОБ № 1 ДЗМ"                 ЦАО    AQUILION_GKOB1   
3  ГБУЗ "ГКОБ № 1 ДЗМ", Филиал № 1                ЮЗАО  AQUILION_GKOB1F1   
4     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   

  DEVICE_TYPE                                          STUDY_UID SERVICE_TYPE  \
0          КТ  1.2.840.113619.2.428.3.1930040987.340.16698629...           КТ   
1          КТ  1.2.840.113619.2.428.3.1930040987.340.16698629...           КТ   
2          КТ  1.2.40.0.13.1.20397098119916846422725270055549...           КТ   
3          КТ  1.2.40.0.13.1.31668921814789388389199005221429...           КТ   
4          КТ  1.2.840.113619.2.428.3.1930040987.288.16696088...           КТ   

   BODY_PART  PROCEDURE_CODE  \
0  WHOLEBODY             404   
1  WHOLEBODY             404   
2  WHOLEBODY             404   
3  WHOLEBODY             404   
4  WHOLEBODY             404   

                                      PROCEDURE_NAME       STUDY_DATETIME  \
0  Компьютерная томография органов грудной клетки...  02.12.2022 13:04:00   
1  Компьютерная томография органов грудной клетки...  02.12.2022 08:38:00   
2  Компьютерная томография органов грудной клетки...  04.12.2022 11:50:00   
3  Компьютерная томография органов грудной клетки...  04.12.2022 09:15:00   
4  Компьютерная томография органов грудной клетки...  30.11.2022 13:13:00   

   ...    LAST_VALIDATION_2 DOCTOR_CODE_1  \
0  ...  02.12.2022 16:43:58     CNTCT1278   
1  ...  03.12.2022 13:26:21     CNTCT1278   
2  ...  04.12.2022 13:45:48     CNTCT1284   
3  ...  04.12.2022 13:47:20     CNTCT1284   
4  ...  30.11.2022 20:44:24      CNTCT116   

                                         DESCRIPTION  \
0  \nПри сравнении с исследованием от 29/09/22 - ...   
1  \nИсследование повторное, предыдущее от 27/08/...   
2  \nСравнение с 18.10.22:\n\nКлинические данные:...   
3  \nСравнение с 22.10.22:\n\nКлинические данные:...   
4  \nВ сравнении с КТ исследованием от 23.08.22г ...   

                                          CONCLUSION PNM_SEVERITY_GDT  \
0  \nТип ответа по recist: нельзя оценить\n\nПри ...              NaN   
1  \nИзмеримых очагов для оценки по Recist 1.1 не...              NaN   
2  \nКТ-картина очаговых изменений легких вторичн...              NaN   
3  \nКТ-картина опухоли левой молочной железы, вт...              NaN   
4  \nКТ-картина центральной опухоли верхней доли ...              NaN   

  CLINTAGS_TUB_VALUE PNM_SEVERITY_ADT RECOMMENDATION keyword  \
0                NaN              NaN            NaN  RECIST   
1                NaN              NaN            NaN  RECIST   
2                NaN              NaN            NaN  RECIST   
3                NaN              NaN            NaN  RECIST   
4                NaN              NaN            NaN  RECIST   

                        estimation_Recist  
0                                          
1                                          
2           стабилизация (Stable Disease)  
3  прогрессирование (Progressive Disease)  
4           стабилизация (Stable Disease)  

[5 rows x 24 columns]

стабилизация (Stable Disease)             63
                                          50
прогрессирование (Progressive Disease)    42
частичный ответ (Partial Response)        20
полный ответ (Complete Response)           4
Name: estimation_Recist, dtype: int64

<span style="color:blue">К пустым 42-м строкам добавились еще 8 строк, всего их стало 50.</span>

In [8]:
# 6. Если в CONCLUSION имеется ключевое слово «Recist», но не указан тип ответа, 
# то в поле estimation_Recist укажем «Ошибка описания»:

for i in range (data2.shape[0]):
    if (
    'recist' in data2['CONCLUSION'][i].lower() 
    and 'стабилизация' not in data2['CONCLUSION'][i].lower() 
    and 'прогрессирование' not in data2['CONCLUSION'][i].lower()
    and 'частичный ответ' not in data2['CONCLUSION'][i].lower()
    and 'полный ответ' not in data2['CONCLUSION'][i].lower()
    ):
        data2['estimation_Recist'][i] = 'ошибка описания'
            
display(data2.head())
data2['estimation_Recist'].value_counts()

ID_MO_EMIAS            DISTRICT          AE_TITLE  \
0     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   
1     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   
2              ГБУЗ "ГКОБ № 1 ДЗМ"                 ЦАО    AQUILION_GKOB1   
3  ГБУЗ "ГКОБ № 1 ДЗМ", Филиал № 1                ЮЗАО  AQUILION_GKOB1F1   
4     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   

  DEVICE_TYPE                                          STUDY_UID SERVICE_TYPE  \
0          КТ  1.2.840.113619.2.428.3.1930040987.340.16698629...           КТ   
1          КТ  1.2.840.113619.2.428.3.1930040987.340.16698629...           КТ   
2          КТ  1.2.40.0.13.1.20397098119916846422725270055549...           КТ   
3          КТ  1.2.40.0.13.1.31668921814789388389199005221429...           КТ   
4          КТ  1.2.840.113619.2.428.3.1930040987.288.16696088...           КТ   

   BODY_PART  PROCEDURE_CODE  \
0  WHOLEBODY             404   
1  WHOLEBODY             404   
2  WHOLEBODY             404   
3  WHOLEBODY             404   
4  WHOLEBODY             404   

                                      PROCEDURE_NAME       STUDY_DATETIME  \
0  Компьютерная томография органов грудной клетки...  02.12.2022 13:04:00   
1  Компьютерная томография органов грудной клетки...  02.12.2022 08:38:00   
2  Компьютерная томография органов грудной клетки...  04.12.2022 11:50:00   
3  Компьютерная томография органов грудной клетки...  04.12.2022 09:15:00   
4  Компьютерная томография органов грудной клетки...  30.11.2022 13:13:00   

   ...    LAST_VALIDATION_2 DOCTOR_CODE_1  \
0  ...  02.12.2022 16:43:58     CNTCT1278   
1  ...  03.12.2022 13:26:21     CNTCT1278   
2  ...  04.12.2022 13:45:48     CNTCT1284   
3  ...  04.12.2022 13:47:20     CNTCT1284   
4  ...  30.11.2022 20:44:24      CNTCT116   

                                         DESCRIPTION  \
0  \nПри сравнении с исследованием от 29/09/22 - ...   
1  \nИсследование повторное, предыдущее от 27/08/...   
2  \nСравнение с 18.10.22:\n\nКлинические данные:...   
3  \nСравнение с 22.10.22:\n\nКлинические данные:...   
4  \nВ сравнении с КТ исследованием от 23.08.22г ...   

                                          CONCLUSION PNM_SEVERITY_GDT  \
0  \nТип ответа по recist: нельзя оценить\n\nПри ...              NaN   
1  \nИзмеримых очагов для оценки по Recist 1.1 не...              NaN   
2  \nКТ-картина очаговых изменений легких вторичн...              NaN   
3  \nКТ-картина опухоли левой молочной железы, вт...              NaN   
4  \nКТ-картина центральной опухоли верхней доли ...              NaN   

  CLINTAGS_TUB_VALUE PNM_SEVERITY_ADT RECOMMENDATION keyword  \
0                NaN              NaN            NaN  RECIST   
1                NaN              NaN            NaN  RECIST   
2                NaN              NaN            NaN  RECIST   
3                NaN              NaN            NaN  RECIST   
4                NaN              NaN            NaN  RECIST   

                        estimation_Recist  
0                         ошибка описания  
1                         ошибка описания  
2           стабилизация (Stable Disease)  
3  прогрессирование (Progressive Disease)  
4           стабилизация (Stable Disease)  

[5 rows x 24 columns]

стабилизация (Stable Disease)             63
ошибка описания                           42
прогрессирование (Progressive Disease)    42
частичный ответ (Partial Response)        20
                                           8
полный ответ (Complete Response)           4
Name: estimation_Recist, dtype: int64

<span style="color:blue">Ошибка описания указана в 42 строках, т.е. в этих строках имеется ключевое слово «Recist», но не указан тип ответа.</span>

In [9]:
# 8. Все ответы в поле «estimation_Recist» должны быть записаны строго в форматах, указанных в п.4 и 6 выше.

In [10]:
# 9. Создадим столбец "Date_conclusion" с датой заключения в формате дд.мм.гггг:

data2['STUDY_DATETIME'] = pd.to_datetime(data2['STUDY_DATETIME'])
data2['LAST_VALIDATION_1'] = pd.to_datetime(data2['LAST_VALIDATION_1'])
data2['Date_conclusion'] = data2['STUDY_DATETIME'].dt.strftime('%d.%m.%Y')
data2.head()

ID_MO_EMIAS            DISTRICT          AE_TITLE  \
0     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   
1     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   
2              ГБУЗ "ГКОБ № 1 ДЗМ"                 ЦАО    AQUILION_GKOB1   
3  ГБУЗ "ГКОБ № 1 ДЗМ", Филиал № 1                ЮЗАО  AQUILION_GKOB1F1   
4     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   

  DEVICE_TYPE                                          STUDY_UID SERVICE_TYPE  \
0          КТ  1.2.840.113619.2.428.3.1930040987.340.16698629...           КТ   
1          КТ  1.2.840.113619.2.428.3.1930040987.340.16698629...           КТ   
2          КТ  1.2.40.0.13.1.20397098119916846422725270055549...           КТ   
3          КТ  1.2.40.0.13.1.31668921814789388389199005221429...           КТ   
4          КТ  1.2.840.113619.2.428.3.1930040987.288.16696088...           КТ   

   BODY_PART  PROCEDURE_CODE  \
0  WHOLEBODY             404   
1  WHOLEBODY             404   
2  WHOLEBODY             404   
3  WHOLEBODY             404   
4  WHOLEBODY             404   

                                      PROCEDURE_NAME      STUDY_DATETIME  ...  \
0  Компьютерная томография органов грудной клетки... 2022-02-12 13:04:00  ...   
1  Компьютерная томография органов грудной клетки... 2022-02-12 08:38:00  ...   
2  Компьютерная томография органов грудной клетки... 2022-04-12 11:50:00  ...   
3  Компьютерная томография органов грудной клетки... 2022-04-12 09:15:00  ...   
4  Компьютерная томография органов грудной клетки... 2022-11-30 13:13:00  ...   

  DOCTOR_CODE_1                                        DESCRIPTION  \
0     CNTCT1278  \nПри сравнении с исследованием от 29/09/22 - ...   
1     CNTCT1278  \nИсследование повторное, предыдущее от 27/08/...   
2     CNTCT1284  \nСравнение с 18.10.22:\n\nКлинические данные:...   
3     CNTCT1284  \nСравнение с 22.10.22:\n\nКлинические данные:...   
4      CNTCT116  \nВ сравнении с КТ исследованием от 23.08.22г ...   

                                          CONCLUSION PNM_SEVERITY_GDT  \
0  \nТип ответа по recist: нельзя оценить\n\nПри ...              NaN   
1  \nИзмеримых очагов для оценки по Recist 1.1 не...              NaN   
2  \nКТ-картина очаговых изменений легких вторичн...              NaN   
3  \nКТ-картина опухоли левой молочной железы, вт...              NaN   
4  \nКТ-картина центральной опухоли верхней доли ...              NaN   

  CLINTAGS_TUB_VALUE PNM_SEVERITY_ADT RECOMMENDATION keyword  \
0                NaN              NaN            NaN  RECIST   
1                NaN              NaN            NaN  RECIST   
2                NaN              NaN            NaN  RECIST   
3                NaN              NaN            NaN  RECIST   
4                NaN              NaN            NaN  RECIST   

                        estimation_Recist Date_conclusion  
0                         ошибка описания      12.02.2022  
1                         ошибка описания      12.02.2022  
2           стабилизация (Stable Disease)      12.04.2022  
3  прогрессирование (Progressive Disease)      12.04.2022  
4           стабилизация (Stable Disease)      30.11.2022  

[5 rows x 25 columns]

<span style="color:blue">Столбец Date_conclusion создан в нужном формате.</span>

In [11]:
# 10. Создадим столбец "Time_till_conclusion" и рассчитаем значение как LAST_VALIDATION_1 – STUDY_DATETIME:

# найдем разницу между датами
data2['Time_till_conclusion'] = data2['LAST_VALIDATION_1'] - data2['STUDY_DATETIME']

# переведем разницу между датами в секунды
data2['Time_till_conclusion'] = data2['Time_till_conclusion'].dt.total_seconds()

# создадим функцию для перевода секунд в формат чч.мм.сс
def convert(sec): 
    sec = sec % (24 * 3600) 
    hour = sec // 3600 
    sec %= 3600 
    min = sec // 60 
    sec %= 60 
    return ("%02d.%02d.%02d" % (hour, min, sec)) 

# применим функцию к столбцу 'Time_till_conclusion'
data2['Time_till_conclusion'] = data2['Time_till_conclusion'].apply(convert)
data2.head()

ID_MO_EMIAS            DISTRICT          AE_TITLE  \
0     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   
1     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   
2              ГБУЗ "ГКОБ № 1 ДЗМ"                 ЦАО    AQUILION_GKOB1   
3  ГБУЗ "ГКОБ № 1 ДЗМ", Филиал № 1                ЮЗАО  AQUILION_GKOB1F1   
4     ГБУЗ "ММКЦ "Коммунарка" ДЗМ"  Московская область     REVOL2_GKB40K   

  DEVICE_TYPE                                          STUDY_UID SERVICE_TYPE  \
0          КТ  1.2.840.113619.2.428.3.1930040987.340.16698629...           КТ   
1          КТ  1.2.840.113619.2.428.3.1930040987.340.16698629...           КТ   
2          КТ  1.2.40.0.13.1.20397098119916846422725270055549...           КТ   
3          КТ  1.2.40.0.13.1.31668921814789388389199005221429...           КТ   
4          КТ  1.2.840.113619.2.428.3.1930040987.288.16696088...           КТ   

   BODY_PART  PROCEDURE_CODE  \
0  WHOLEBODY             404   
1  WHOLEBODY             404   
2  WHOLEBODY             404   
3  WHOLEBODY             404   
4  WHOLEBODY             404   

                                      PROCEDURE_NAME      STUDY_DATETIME  ...  \
0  Компьютерная томография органов грудной клетки... 2022-02-12 13:04:00  ...   
1  Компьютерная томография органов грудной клетки... 2022-02-12 08:38:00  ...   
2  Компьютерная томография органов грудной клетки... 2022-04-12 11:50:00  ...   
3  Компьютерная томография органов грудной клетки... 2022-04-12 09:15:00  ...   
4  Компьютерная томография органов грудной клетки... 2022-11-30 13:13:00  ...   

                                         DESCRIPTION  \
0  \nПри сравнении с исследованием от 29/09/22 - ...   
1  \nИсследование повторное, предыдущее от 27/08/...   
2  \nСравнение с 18.10.22:\n\nКлинические данные:...   
3  \nСравнение с 22.10.22:\n\nКлинические данные:...   
4  \nВ сравнении с КТ исследованием от 23.08.22г ...   

                                          CONCLUSION PNM_SEVERITY_GDT  \
0  \nТип ответа по recist: нельзя оценить\n\nПри ...              NaN   
1  \nИзмеримых очагов для оценки по Recist 1.1 не...              NaN   
2  \nКТ-картина очаговых изменений легких вторичн...              NaN   
3  \nКТ-картина опухоли левой молочной железы, вт...              NaN   
4  \nКТ-картина центральной опухоли верхней доли ...              NaN   

  CLINTAGS_TUB_VALUE PNM_SEVERITY_ADT RECOMMENDATION keyword  \
0                NaN              NaN            NaN  RECIST   
1                NaN              NaN            NaN  RECIST   
2                NaN              NaN            NaN  RECIST   
3                NaN              NaN            NaN  RECIST   
4                NaN              NaN            NaN  RECIST   

                        estimation_Recist Date_conclusion Time_till_conclusion  
0                         ошибка описания      12.02.2022             03.39.58  
1                         ошибка описания      12.02.2022             03.28.38  
2           стабилизация (Stable Disease)      12.04.2022             01.55.48  
3  прогрессирование (Progressive Disease)      12.04.2022             01.59.33  
4           стабилизация (Stable Disease)      30.11.2022             07.31.24  

[5 rows x 26 columns]

<span style="color:blue">Столбец Time_till_conclusion создан в нужном формате.</span>

In [12]:
# Выгрузим финальный датафрейм в файл excel:

#data2.to_excel(r'/mnt/HC_Volume_18315164/home-jupyter/jupyter-d-kirilovichev/gbuz.xlsx')